cpp/src/endpoint.cpp

```diff
@@ -234,27 +225,13 @@ size_t Endpoint::cancelInflightRequests()
   size_t canceled = 0;
 
   if (worker->isProgressThreadRunning()) {
-    auto statusMutex             = std::make_shared<std::mutex>();
-    auto statusConditionVariable = std::make_shared<std::condition_variable>();
-    auto completed               = std::make_shared<bool>(false);
-
-    auto setterCancel = [this, &canceled, completed]() {
-      canceled   = _inflightRequests->cancelAll();
-      *completed = true;
-    };
-    auto getter = [completed]() { return *completed; };
-
-    worker->registerGenericPre([&statusMutex, &statusConditionVariable, &setterCancel]() {
-      ucxx::utils::conditionSetter(statusMutex, statusConditionVariable, setterCancel);
-    });
-    ucxx::utils::conditionGetter(statusMutex, statusConditionVariable, completed, getter);
-
-    *completed        = false;
-    auto setterVerify = [this, completed]() { *completed = true; };
-    worker->registerGenericPost([&statusMutex, &statusConditionVariable, &setterVerify]() {
-      ucxx::utils::conditionSetter(statusMutex, statusConditionVariable, setterVerify);
-    });
-    ucxx::utils::conditionGetter(statusMutex, statusConditionVariable, completed, getter);
+    utils::CallbackNotifier callback_pre{
+      false, [this, &canceled]() { canceled = _inflightRequests->cancelAll(); }};
```

Comment:
In this case, we call the extra function with the lock held (and write to the shared `canceled` output variable.

If we want more uniform semantics we could implement the notifier like so:

```c++
template <typename T>
struct Notifier {
  /**
   * @brief Construct a thread-safe notification object with given initial value
   *
   * @param[in] init   The initial flag value
   */
  Notifier(T init) : _result{init} {};
  Notifier(const Notifier &) = delete;
  Notifier &operator=(Notifier const &) = delete;
  Notifier(Notifier &&o) = delete;
  Notifier &operator=(Notifier &&o) = delete;

  /**
   * @brief Store a new value and notify a waiting thread
   *
   * @param[in] fn   Function of type void -> T called with lock held
   *                 to produce new value.
   */
  void store(auto fn) {
    {
      std::lock_guard lock(_m);
      _result = fn();
    }
    _cv.notify_one();
  }
  /**
   * @brief Wait while predicate is not true for the flag value to change
   *
   * @param[in] p   Function of type T -> bool called with the flag
   *                value. This function loops until the predicate is
   *                satisfied. See also std::condition_variable::wait
   * @param[out]    New flag value
   */
  template <typename P> T wait(P p) {
    std::unique_lock lock(_m);
    _cv.wait(lock, [this, p]() { return p(_result); });
    return std::move(_result);
  }

private:
  T _result;                    //< flag to spin on
  std::mutex _m;                //< lock to guard accesses
  std::condition_variable _cv;  //< notification condition var
};
```

And then use like so:
```c++
  using T = std::pair<bool, size_t>;
  Notifier<T> status{{false, 0}};
  worker->registerGenericPre([&status]() {
    status.store([this]() -> T {
      return {true, _inflightRequests->cancelAll()};
    });
  });
  auto [_, canceled] = status.wait([](auto s) { return s.first; });
```

But perhaps that is a bit ugly.

Reply:
Is there a reason that you need this function call at all, rather than just doing the `_canceled = ...` in the `registerGenericPre` callback.

Reply:
I see where you're going with the uniform semantics post, I wouldn't necessarily call the code uglybut I think it's a little bit more complex to read. Should we keep it as is for now?

> Is there a reason that you need this function call at all, rather than just doing the `_canceled = ...` in the `registerGenericPre` callback.

I'm not sure I understand which function call you're referring to. There's no `_canceled` object, we only have `canceled` as a local variable as that result will change for each call. The `_inflightRequests->cancelAll()` must be called, as that is what will indeed trigger cancelation of requests.

Reply:
```
[this, &canceled]() { canceled = _inflightRequests->cancelAll(); }
```

This one (which runs on the worker progress thread with the lock held), could be just called in the `registergenericpre` callback before calling `status.store`?

Reply:
> I see where you're going with the uniform semantics post, I wouldn't necessarily call the code uglybut I think it's a little bit more complex to read. Should we keep it as is for now?

Yeah, if this were haskell and anonymous/partially applied functions weren't so ugly, everyone would be used to it, but double nesting is probably a C++ antipattern.

Reply:
> This one (which runs on the worker progress thread with the lock held), could be just called in the `registergenericpre` callback before calling `status.store`?

Yes, that's a good suggestion, simplified the code a bit in https://github.com/rapidsai/ucxx/pull/65/commits/b5548689b3ae04f62bdb02ce13d6895e5ea74f6f .